physx/src/visual_debugger.rs

```diff
@@ -23,104 +27,155 @@ pub enum VisualDebuggerSceneFlag {
+unsafe impl Class<physx_sys::PxPvd> for VisualDebugger {
+    fn as_ptr(&self) -> *const physx_sys::PxPvd {
+        self.pvd.as_ptr()
     }
 
-    /// Connect the visual debugger over the provided transport. Returns `true`
-    /// if the connection succeded.
-    fn connect(
-        &mut self,
-        transport: *mut PxPvdTransport,
-        flags: PxPvdInstrumentationFlags,
-    ) -> bool {
-        unsafe { PxPvd_connect_mut(self.get_raw_mut(), transport, flags) }
-    }
-}
-
-#[physx_type]
-impl PvdSceneClient {
-    /// Enable and disable what should be transmitted to the visual debugger
-    /// instance.
-    pub fn set_scene_flags(&mut self, flags: BitFlags<VisualDebuggerSceneFlag>) {
-        unsafe {
-            PxPvdSceneClient_setScenePvdFlags_mut(
-                self.ptr,
-                PxPvdSceneFlags {
-                    mBits: flags.bits(),
-                },
-            )
-        }
+    fn as_mut_ptr(&mut self) -> *mut physx_sys::PxPvd {
+        self.pvd.as_mut_ptr()
     }
 }
 
 impl VisualDebugger {
-    /// Create a new VisualDebugger instance. This is a utility class that
-    /// combines the TCP setup and the actual Pvd into one object. The port
+    /// Create a new VisualDebugger instance, a utility class that
+    /// combines the TCP setup and the Pvd into one object. The port
     /// default for the PVD program is port 5425, so it is suggested to use this
     /// unless you're explicitly changing the other one as well.
     ///
     /// This function internally calls `new_with_timeout` with a default timeout
     /// of 10 ms.
-    pub fn new(foundation: &mut Foundation, port: i32) -> Self {
+    pub fn new(foundation: &mut Foundation, port: i32) -> Option<Self> {
         VisualDebugger::new_with_timeout(foundation, port, 10)
     }
 
     /// See description of `new`
-    pub fn new_with_timeout(foundation: &mut Foundation, port: i32, timeout: u32) -> Self {
+    pub fn new_with_timeout(foundation: &mut Foundation, port: i32, timeout: u32) -> Option<Self> {
         use std::ffi::CStr;
 
         let flags = PxPvdInstrumentationFlags {
             mBits: PxPvdInstrumentationFlag::eDEBUG as u8,
         };
 
-        let mut pvd = Pvd::new(foundation);
+        let pvd = unsafe { Pvd::from_raw(phys_PxCreatePvd(foundation.as_mut_ptr()))? };
         let transport = unsafe {
             let oshost = CStr::from_bytes_with_nul_unchecked(b"localhost\0");
-            phys_PxDefaultPvdSocketTransportCreate(oshost.as_ptr() as _, port, timeout)
+            PvdTransport::from_raw(phys_PxDefaultPvdSocketTransportCreate(
+                oshost.as_ptr() as _,
+                port,
+                timeout,
+            ))?
         };
 
-        if !pvd.connect(transport, flags) {
-            warn!("failed connecting visual debugger, running in no-debug mode");
-        }
+        let mut visual_debugger = Self { pvd, transport };
 
-        Self { pvd, transport }
+        visual_debugger.connect(flags);
```

Comment:
I changed this so that the PvdTransport is wrapped in an Option, and that is None if connection fails.  I think reconnecting while running is supported, and certainly launching without connecting it right away should be. I added a `set_transport` method and a couple more methods from the API to support this.